File: TimeKeeperAndroid/src/com/tjoris/timekeeper/data/FilePlaylistStore.java
package com.tjoris.timekeeper.data;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import org.xmlpull.v1.XmlSerializer;

import android.os.Environment;
import android.util.Log;
import android.util.Xml;

public class FilePlaylistStore implements IPlaylistStore
{
	private static final String kTAG_PLAYLIST = "playlist";
	private static final String kTAG_SONG = "song";
	private static final String kATTR_NAME = "name";
	private static final String kATTR_WEIGHT = "weight";
	private static final String kATTR_TEMPO = "tempo";
	
	
	private static class EntryComparator implements Comparator<AbstractEntry>
	{
		@Override
		public int compare(final AbstractEntry lhs, final AbstractEntry rhs)
		{
			return lhs.getWeight() - rhs.getWeight();
		}
	}
	
	private static final File kBASE_PATH = new File(Environment.getExternalStorageDirectory(), "TimeKeeper");
	private static final EntryComparator kCOMPARATOR = new EntryComparator();
	private static final Pattern kFILENAME_PATTERN = Pattern.compile("([0-9]+)\\.xml");
	
	/*private final SQLPlaylistStore fSQLStore;
	
	
	FilePlaylistStore(final SQLPlaylistStore sqlStore)
    {
		fSQLStore = sqlStore;
    */
	
	@Override
	public List<PlaylistHeader> readAllPlaylists()
	{
		/*if (kBASE_PATH.listFiles().length == 1)
		{
			for (final PlaylistHeader playlistHeader : fSQLStore.readAllPlaylists())
			{
				final Playlist playlist = fSQLStore.readPlaylist(playlistHeader.getID());
				playlist.setID(AbstractEntry.kID_NEW);
				addPlaylist(playlist);
			}
		}*/

		final List<PlaylistHeader> playlists = new ArrayList<PlaylistHeader>();
		for (final File file : kBASE_PATH.listFiles())
		{
			final Matcher matcher = kFILENAME_PATTERN.matcher(file.getName());
			if (matcher.matches())
			{
				final PlaylistHeader playlistHeader = readPlaylistHeader(file, matcher.group(1));
				if (playlistHeader != null)
				{
					playlists.add(playlistHeader);
				}
			}
		}
		Collections.sort(playlists, kCOMPARATOR);
	    return playlists;
	}
	
	@Override
	public void storePlaylistHeader(final PlaylistHeader playlist)
	{
		final Playlist completePlaylist = readPlaylist(playlist.getID());
		completePlaylist.setName(playlist.getName());
		completePlaylist.setWeight(playlist.getWeight());
		savePlaylist(completePlaylist);
	}
	
	public Playlist readPlaylist(final long id)
	{
		try
		{
			final Playlist playlist = new Playlist(id, null, 0);
			final InputStream in = new BufferedInputStream(new FileInputStream(getPlaylistFile(playlist)));
			try
			{
				final XmlPullParser parser = Xml.newPullParser();
				parser.setInput(in, "UTF-8");
				for (;;)
				{
					switch (parser.next())
					{
					case XmlPullParser.START_TAG:
					{
						if (kTAG_PLAYLIST.equals(parser.getName()))
						{
							final String name = parser.getAttributeValue(null, kATTR_NAME);
							final int weight = Integer.parseInt(parser.getAttributeValue(null, kATTR_WEIGHT));
							playlist.setName(name);
							playlist.setWeight(weight);
						}
						else if (kTAG_SONG.equals(parser.getName()))
						{
							final String name = parser.getAttributeValue(null, kATTR_NAME);
							final int tempo = Integer.parseInt(parser.getAttributeValue(null, kATTR_TEMPO));
							playlist.addSong(new Song(name, tempo));
						}
						break;
					}
					case XmlPullParser.END_DOCUMENT:
					{
						return playlist;
					}
					default:
					{
						break;
					}
					}
				}
			}
			finally
			{
				in.close();
			}
		}
		catch (final XmlPullParserException | IOException e)
		{
			Log.e("TimeKeeper", "Could not read playlist: " + e.getMessage(), e);
		}
		return null;
	}
	
	@Override
	public void addPlaylist(final Playlist playlist)
	{
		savePlaylist(playlist);
	}
	
	@Override
	public void deletePlaylist(final PlaylistHeader playlist)
	{
		getPlaylistFile(playlist).delete();
	}
	
	@Override
	public int getNextPlaylistWeight()
	{
		final List<PlaylistHeader> playlists = readAllPlaylists();
		if (playlists.isEmpty())
		{
			return 0;
		}
		return playlists.get(playlists.size() - 1).getWeight() + 1;
	}
	
	@Override
	public void storeSong(final Playlist playlist, final Song song)
	{
		savePlaylist(playlist);
	}
	
	@Override
	public void deleteSong(final Playlist playlist, final Song song)
	{
		savePlaylist(playlist);
	}
	
	@Override
	public void close()
	{
	}
	
	private PlaylistHeader readPlaylistHeader(final File file, final String filename)
	{
		try
		{
			final long id = Long.parseLong(filename);
			final InputStream in = new BufferedInputStream(new FileInputStream(file));
			try
			{
				final XmlPullParser parser = Xml.newPullParser();
				parser.setInput(in, "UTF-8");
				for (;;)
				{
					switch (parser.next())
					{
					case XmlPullParser.START_TAG:
					{
						parser.require(XmlPullParser.START_TAG, null, kTAG_PLAYLIST);
						final String name = parser.getAttributeValue(null, kATTR_NAME);
						final int weight = Integer.parseInt(parser.getAttributeValue(null, kATTR_WEIGHT));
						return new PlaylistHeader(id, name, weight);
					}
					case XmlPullParser.END_DOCUMENT:
					{
						return null;
					}
					default:
					{
						break;
					}
					}
				}
			}
			finally
			{
				in.close();
			}
		}
		catch (final XmlPullParserException | IOException e)
		{
			Log.e("TimeKeeper", "Could not read playlist: " + e.getMessage(), e);
		}
		return null;
	}
	
	private void savePlaylist(final Playlist playlist)
	{
		try
		{
			if (playlist.isNew())
			{
				playlist.setID(System.currentTimeMillis());
			}
			final XmlSerializer serializer = Xml.newSerializer();
			final OutputStream out = new BufferedOutputStream(new FileOutputStream(getPlaylistFile(playlist)));
			try
			{
				serializer.setOutput(out, "UTF-8");
				serializer.startDocument("UTF-8", null);
				serializer.startTag(null, kTAG_PLAYLIST);
				serializer.attribute(null, kATTR_NAME, playlist.getName());
				serializer.attribute(null, kATTR_WEIGHT, Integer.toString(playlist.getWeight()));
				for (final Song song : playlist.getSongs())
				{
					serializer.startTag(null, kTAG_SONG);
					serializer.attribute(null, kATTR_NAME, song.getName());
					serializer.attribute(null, kATTR_TEMPO, Integer.toString(song.getTempo()));
					serializer.endTag(null, kTAG_SONG);
				}
				serializer.endTag(null, kTAG_PLAYLIST);
				serializer.endDocument();
			}
			finally
			{
				out.close();
			}
		}
		catch (final IOException e)
		{
			Log.e("TimeKeeper", "Could not write playlist: " + e.getMessage(), e);
		}
	}
	
	private File getPlaylistFile(final PlaylistHeader playlist)
	{
		return new File(kBASE_PATH, Long.toString(playlist.getID()) + ".xml");
	}
}

File: TimeKeeperAndroid/app/src/main/java/be/t_ars/timekeeper/data/FilePlaylistStore.kt
package be.t_ars.timekeeper.data

import android.os.Environment
import android.util.Log
import android.util.Xml
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.*
import java.util.*
import java.util.regex.Pattern

class FilePlaylistStore : IPlaylistStore {
    private class EntryComparator : Comparator<AbstractEntry> {
        override fun compare(lhs: AbstractEntry, rhs: AbstractEntry): Int {
            return lhs.weight - rhs.weight
        }
    }

    override fun readAllPlaylists(): List<PlaylistHeader> {
        val playlists = ArrayList<PlaylistHeader>()
        val files = kBASE_PATH.listFiles()
        files?.forEach { file ->
            val matcher = kFILENAME_PATTERN.matcher(file.name)
            if (matcher.matches()) {
                val id = matcher.group(1)
                readPlaylistHeader(file, id)?.let { playlists.add(it) }
            }
        }
        Collections.sort(playlists, kCOMPARATOR)
        return playlists
    }

    override fun storePlaylistHeader(playlist: PlaylistHeader) {
        readPlaylist(playlist.id)?.let { completePlaylist ->
            completePlaylist.name = playlist.name
            completePlaylist.weight = playlist.weight
            savePlaylist(completePlaylist)
        }
    }

    override fun readPlaylist(id: Long): Playlist? {
        try {
            var playlist: Playlist? = null
            BufferedInputStream(FileInputStream(getPlaylistFile(id))).use { inputStream ->
                val parser = Xml.newPullParser()
                parser.setInput(inputStream, "UTF-8")
                while (true) {
                    when (parser.next()) {
                        XmlPullParser.START_TAG -> {
                            if (kTAG_PLAYLIST == parser.name) {
                                val name = parser.getAttributeValue(null, kATTR_NAME)
                                val weight = Integer.parseInt(parser.getAttributeValue(null, kATTR_WEIGHT))
                                playlist = Playlist(id, name, weight)
                            } else if (kTAG_SONG == parser.name) {
                                val name = parser.getAttributeValue(null, kATTR_NAME)
                                val tempo = parser.getAttributeValue(null, kATTR_TEMPO)
                                        ?.let(Integer::parseInt)
                                playlist?.addSong(Song(name = name, tempo = tempo))
                            }
                        }
                        XmlPullParser.END_DOCUMENT -> {
                            return playlist
                        }
                        else -> {
                        }
                    }
                }
            }
        } catch (e: XmlPullParserException) {
            Log.e("TimeKeeper", "Could not read playlist: " + e.message, e)
        } catch (e: IOException) {
            Log.e("TimeKeeper", "Could not read playlist: " + e.message, e)
        }

        return null
    }

    override fun addPlaylist(playlist: Playlist) {
        savePlaylist(playlist)
    }

    override fun deletePlaylist(playlist: PlaylistHeader) {
        getPlaylistFile(playlist.id).delete()
    }

    override val nextPlaylistWeight: Int
        get() {
            val playlists = readAllPlaylists()
            if (playlists.isEmpty()) {
                return 0
            }
            return playlists[playlists.size - 1].weight + 1
        }

    override fun storeSong(playlist: Playlist, song: Song) {
        savePlaylist(playlist)
    }

    override fun deleteSong(playlist: Playlist, song: Song) {
        savePlaylist(playlist)
    }

    override fun close() {}

    private fun readPlaylistHeader(file: File, filename: String): PlaylistHeader? {
        try {
            val id = java.lang.Long.parseLong(filename)
            BufferedInputStream(FileInputStream(file)).use { stream ->
                val parser = Xml.newPullParser()
                parser.setInput(stream, "UTF-8")
                while (true) {
                    when (parser.next()) {
                        XmlPullParser.START_TAG -> {
                            parser.require(XmlPullParser.START_TAG, null, kTAG_PLAYLIST)
                            val name = parser.getAttributeValue(null, kATTR_NAME)
                            val weight = Integer.parseInt(parser.getAttributeValue(null, kATTR_WEIGHT))
                            return PlaylistHeader(id, name, weight)
                        }
                        XmlPullParser.END_DOCUMENT -> {
                            return null
                        }
                        else -> {
                        }
                    }
                }
            }
        } catch (e: XmlPullParserException) {
            Log.e("TimeKeeper", "Could not read playlist: " + e.message, e)
        } catch (e: IOException) {
            Log.e("TimeKeeper", "Could not read playlist: " + e.message, e)
        }

        return null
    }

    private fun savePlaylist(playlist: Playlist) {
        try {
            kBASE_PATH.mkdirs()
            if (playlist.isNew()) {
                playlist.id = System.currentTimeMillis()
            }
            val serializer = Xml.newSerializer()
            BufferedOutputStream(FileOutputStream(getPlaylistFile(playlist.id))).use { outputStream ->
                serializer.setOutput(outputStream, "UTF-8")
                serializer.startDocument("UTF-8", null)
                serializer.startTag(null, kTAG_PLAYLIST)
                serializer.attribute(null, kATTR_NAME, playlist.name)
                serializer.attribute(null, kATTR_WEIGHT, playlist.weight.toString())
                for (song in playlist.songs) {
                    serializer.startTag(null, kTAG_SONG)
                    serializer.attribute(null, kATTR_NAME, song.name)
                    if (song.tempo != null)
                        serializer.attribute(null, kATTR_TEMPO, song.tempo.toString())
                    serializer.endTag(null, kTAG_SONG)
                }
                serializer.endTag(null, kTAG_PLAYLIST)
                serializer.endDocument()
            }
        } catch (e: IOException) {
            Log.e("TimeKeeper", "Could not write playlist: " + e.message, e)
        }

    }

    private fun getPlaylistFile(id: Long): File {
        return File(kBASE_PATH, "$id.xml")
    }

    companion object {
        private const val kTAG_PLAYLIST = "playlist"
        private const val kTAG_SONG = "song"
        private const val kATTR_NAME = "name"
        private const val kATTR_WEIGHT = "weight"
        private const val kATTR_TEMPO = "tempo"

        private val kBASE_PATH = File(Environment.getExternalStorageDirectory(), "TimeKeeper")
        private val kCOMPARATOR = EntryComparator()
        private val kFILENAME_PATTERN = Pattern.compile("([0-9]+)\\.xml")
    }
}
